model/task/db.go

```diff
@@ -2643,17 +2662,60 @@ func getTasksByVersionPipeline(versionID string, opts GetTasksByVersionOptions)
 	return pipeline, nil
 }
 
-func (t *Task) FindAllUnmarkedBlockedDependencies() ([]Task, error) {
-	okStatusSet := []string{AllStatuses, t.Status}
-	query := db.Query(bson.M{
-		DependsOnKey: bson.M{"$elemMatch": bson.M{
-			DependencyTaskIdKey:       t.Id,
-			DependencyStatusKey:       bson.M{"$nin": okStatusSet},
-			DependencyUnattainableKey: false,
-		},
-		}},
-	)
-	return FindAll(query)
+// FindAllUnmarkedDependenciesToBlock finds tasks that depend on one of the
+// given tasks. For each task dependency, it finds those that have not been
+// marked unattainable and currently have a status that would block the task
+// from running.
+// This must find tasks in smaller chunks to avoid the 16 MB query size limit -
+// if the number of tasks is large, a single query could be too large and the DB
+// will reject it.
+func FindAllUnmarkedDependenciesToBlock(tasks []Task) ([]Task, error) {
+	if len(tasks) == 0 {
+		return nil, nil
+	}
+
+	const maxTasksPerQuery = 2000
```

Comment:
Do we know what the estimated update size is with 2000 task ids?

Reply:
I estimated that for this query, it's about 1 MB of string text per 1k task IDs (the update could be a lot higher since it has both tasks and dependency IDs). It's on the conservative side to make sure that this query doesn't fail due to the batch size.